pkg/net/bitswap.go

```diff
@@ -187,58 +193,90 @@ func (c BitswapClient) SpadeTraversal(parent context.Context,
 	startTime := time.Now()
 
 	i := uint(0)
+	var blk blocks.Block
 	for {
-		// Retrieval
-		logger.Infof("retrieving %s\n", cidToRetrieve.String())
-		blk, err := c.RetrieveBlock(parent, target, network, bswap, cidToRetrieve)
-
-		if err != nil {
-			return task.NewErrorRetrievalResultWithErrorResolution(task.RetrievalFailure, err), nil
-		}
-
-		// Verify returned content hashes to the CID we're expecting
-		if !blk.Cid().Equals(cidToRetrieve) {
-			return task.NewErrorRetrievalResult(task.CIDMismatch,
-				errors.Errorf("retrieved cid does not match requested: %s, %s",
-					blk.Cid().String(), cidToRetrieve)), nil
+		for j, cidToRetrieve := range cidsToRetrieve {
+			// Retrieval
+			logger.Infof("retrieving %s\n", cidToRetrieve.String())
+			blk, err := c.RetrieveBlock(parent, target, network, bswap, cidToRetrieve)
+
+			if err != nil {
+				return task.NewErrorRetrievalResultWithErrorResolution(task.RetrievalFailure, err), nil
+			}
+
+			// Verify returned content hashes to the CID we're expecting
+			if !blk.Cid().Equals(cidToRetrieve) {
+				return task.NewErrorRetrievalResult(task.CIDMismatch,
+					errors.Errorf("retrieved cid does not match requested: %s, %s",
+						blk.Cid().String(), cidToRetrieve)), nil
+			}
+
+			// Wait until we are at max depth and tried all the challenges
+			if i == maxTraverseDepth && j == len(cidsToRetrieve)-1 {
+				var size = int64(len(blk.RawData()))
+				elapsed := time.Since(startTime)
+				logger.With("size", size).With("elapsed", elapsed).Info("Retrieved block")
+
+				// we've reached the requested depth of the tree
+				return task.NewSuccessfulRetrievalResult(elapsed, size, elapsed), nil
+			}
+
+			// if not at bottom of the tree, keep going down the links until we reach it or hit a dead end
+			links, err := FindLinks(parent, blk)
+			if err != nil {
+				return task.NewErrorRetrievalResultWithErrorResolution(task.CannotDecodeLinks, err), nil
+			}
+
+			logger.Debugf("cid %s has %d links\n", cidToRetrieve.String(), len(links))
+
+			nextIndexes = append(nextIndexes, links...)
```

Comment:
reading through it, the `nextIndexes` variable name was confusing me here... 

maybe`nextLayerIndexes`  would help a bit. but they are just "candidate" indexes , as not every single one of them will be traversed down, just the max. number as specified by the env var

Reply:
I like that. I didnt have a great name for it